From f198f39f37dce73aedb9655daafca098e9e13e3c Mon Sep 17 00:00:00 2001
From: George Kiagiadakis <george.kiagiadakis@collabora.com>
Date: Thu, 12 Jun 2025 07:40:21 +0300
Subject: [PATCH 2/2] lib: module: clear the impl_module pointer when it is
 destroyed by itself

Some modules (like module-loopback) may destroy themselves with
pw_impl_module_schedule_destroy() when the pipewire connection closes.
In that case, we need to clear our pointer so that we don't try to
destroy them again (and crash)

Fixes: #812
---
 lib/wp/module.c | 25 +++++++++++++++++++++----
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/lib/wp/module.c b/lib/wp/module.c
index 11a000c9..0b7ff680 100644
--- a/lib/wp/module.c
+++ b/lib/wp/module.c
@@ -33,6 +33,7 @@ struct _WpImplModule
   WpProperties *props; /* only used during module load */
 
   struct pw_impl_module *pw_impl_module;
+  struct spa_hook impl_module_listener;
 };
 
 G_DEFINE_TYPE (WpImplModule, wp_impl_module, G_TYPE_OBJECT);
@@ -46,6 +47,17 @@ enum {
   PROP_PW_IMPL_MODULE,
 };
 
+static void impl_module_free (void *data)
+{
+  WpImplModule *self = WP_IMPL_MODULE (data);
+  self->pw_impl_module = NULL;
+}
+
+static const struct pw_impl_module_events impl_module_events = {
+  PW_VERSION_IMPL_MODULE_EVENTS,
+  .free = impl_module_free,
+};
+
 static void
 wp_impl_module_init (WpImplModule * self)
 {
@@ -80,10 +92,15 @@ wp_impl_module_constructed (GObject * object)
   self->pw_impl_module =
     pw_context_load_module (context, self->name, self->args, props);
 
-  if (self->pw_impl_module && self->props) {
-    /* With the module loaded, properties are just passthrough now */
-    wp_properties_unref (self->props);
-    self->props = NULL;
+  if (self->pw_impl_module) {
+    if (self->props) {
+      /* With the module loaded, properties are just passthrough now */
+      wp_properties_unref (self->props);
+      self->props = NULL;
+    }
+
+    pw_impl_module_add_listener (self->pw_impl_module,
+        &self->impl_module_listener, &impl_module_events, self);
   }
 
   G_OBJECT_CLASS (wp_impl_module_parent_class)->constructed (object);
-- 
2.52.0

